src/rewritable_units/text_encoder.rs

```diff
@@ -0,0 +1,180 @@
+pub struct StreamingHandlerSink<'output_handler> {
+    non_utf8_encoder: Option<TextEncoder>,
+
+    /// ```compile_fail
+    /// use lol_html::html_content::StreamingHandlerSink;
+    /// struct IsSend<T: Send>(T);
+    /// let x: IsSend<StreamingHandlerSink<'static>>;
+    /// ```
+    ///
+    /// ```compile_fail
+    /// use lol_html::html_content::StreamingHandlerSink;
+    /// struct IsSync<T: Sync>(T);
+    /// let x: IsSync<StreamingHandlerSink<'static>>;
+    /// ```
+    output_handler: &'output_handler mut dyn FnMut(&[u8]),
+}
+
+impl<'output_handler> StreamingHandlerSink<'output_handler> {
+    #[inline(always)]
+    pub(crate) fn new(
+        encoding: &'static Encoding,
+        output_handler: &'output_handler mut dyn FnMut(&[u8]),
+    ) -> Self {
+        Self {
+            non_utf8_encoder: (encoding != UTF_8).then(|| TextEncoder::new(encoding)),
+            output_handler,
+        }
+    }
+
+    /// Writes the given UTF-8 string to the output, converting the encoding and [escaping](ContentType) if necessary.
+    ///
+    /// It may be called multiple times. The strings will be concatenated together.
+    #[inline]
+    pub fn write_str(&mut self, content: &str, content_type: ContentType) {
+        match content_type {
+            ContentType::Html => self.write_html(content),
+            ContentType::Text => self.write_body_text(content),
+        }
+    }
+
+    pub(crate) fn write_html(&mut self, html: &str) {
+        if let Some(encoder) = &mut self.non_utf8_encoder {
+            encoder.encode(html, self.output_handler);
+        } else if !html.is_empty() {
+            (self.output_handler)(html.as_bytes());
+        }
+    }
+
+    /// For text content, not attributes
+    pub(crate) fn write_body_text(&mut self, plaintext: &str) {
+        if let Some(encoder) = &mut self.non_utf8_encoder {
+            escape_body_text(plaintext, &mut |chunk| {
+                debug_assert!(!chunk.is_empty());
+                encoder.encode(chunk, self.output_handler);
+            });
+        } else {
+            escape_body_text(plaintext, &mut |chunk| {
+                debug_assert!(!chunk.is_empty());
+                (self.output_handler)(chunk.as_bytes());
+            });
+        }
+    }
+
+    #[inline]
+    pub(crate) fn output_handler(&mut self) -> &mut dyn FnMut(&[u8]) {
+        &mut self.output_handler
+    }
+}
+
+enum Buffer {
+    Heap(Vec<u8>),
+    Stack([u8; 63]), // leave a byte for the tag
+}
+
+struct TextEncoder {
+    encoder: Encoder,
+    buffer: Buffer,
+}
+
+impl TextEncoder {
+    #[inline]
+    pub fn new(encoding: &'static Encoding) -> Self {
+        debug_assert!(encoding != UTF_8);
+        debug_assert!(encoding.is_ascii_compatible());
+        Self {
+            encoder: encoding.new_encoder(),
+            buffer: Buffer::Stack([0; 63]),
+        }
+    }
+
+    #[inline(never)]
+    fn encode(&mut self, mut content: &str, output_handler: &mut dyn FnMut(&[u8])) {
```

Comment:
I'm confused. This just encodes a Rust string in non-UTF8 byte sequence (if streamed content is non-UTF8) and just writes it to the output sink. Is that right? If so, why not just use the previous impl: https://github.com/cloudflare/lol-html/blob/master/src/base/bytes.rs#L22?

Reply:
1. `Encoding::encode()` contains UB (makes a slice from uninitialized bytes)
2. No-copy fast path is only for UTF-8 to UTF-8 and pure ASCII.
3. If there is any non-ASCII byte, it heap allocates a buffer for the entire string, and uses maximally pessimistic estimate for the buffer size, which is a 3x multiplier for CJK encodings.

My implementation doesn't have UB. It doesn't allocate unless necessary, and can reuse the buffer across multiple chunks.

Reply:
Sounds good. Could you add a code comment explaining all that for the future generations. I imagine it will become lost knowledge otherwise